duckdb-vortex/src/expr/expr.cpp

```diff
@@ -0,0 +1,225 @@
+		dtype->mutable_null();
+		return dtype;
+	case LogicalTypeId::BOOLEAN:
+		dtype->mutable_binary()->set_nullable(nullable);
+		return dtype;
+	case LogicalTypeId::TINYINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::I8);
+		return dtype;
+	case LogicalTypeId::SMALLINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::I16);
+		return dtype;
+	case LogicalTypeId::INTEGER:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::I32);
+		return dtype;
+	case LogicalTypeId::BIGINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::I64);
+		return dtype;
+	case LogicalTypeId::UTINYINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::U8);
+		return dtype;
+	case LogicalTypeId::USMALLINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::U16);
+		return dtype;
+	case LogicalTypeId::UINTEGER:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::U32);
+		return dtype;
+	case LogicalTypeId::UBIGINT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::U64);
+		return dtype;
+	case LogicalTypeId::FLOAT:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::F32);
+		return dtype;
+	case LogicalTypeId::DOUBLE:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::F64);
+		return dtype;
+	case LogicalTypeId::CHAR:
+	case LogicalTypeId::VARCHAR:
+		dtype->mutable_utf8()->set_nullable(nullable);
+		return dtype;
+	case LogicalTypeId::BLOB:
+		dtype->mutable_binary()->set_nullable(nullable);
+		return dtype;
+	default:
+		throw Exception(ExceptionType::NOT_IMPLEMENTED, "into_vortex_dtype", {{"id", type_.ToString()}});
+	}
+}
+
+vortex::scalar::Scalar *into_null_scalar(Arena &arena, LogicalType &logical_type) {
+	auto scalar = Arena::Create<vortex::scalar::Scalar>(&arena);
+	scalar->set_allocated_dtype(into_vortex_dtype(arena, logical_type, true));
+	scalar->mutable_value()->set_null_value(google::protobuf::NULL_VALUE);
+	return scalar;
+}
+
+vortex::scalar::Scalar *into_vortex_scalar(Arena &arena, Value &value, bool nullable) {
+	auto scalar = Arena::Create<vortex::scalar::Scalar>(&arena);
+	auto dtype = into_vortex_dtype(arena, value.type().id(), nullable);
+	scalar->set_allocated_dtype(dtype);
+
+	switch (value.type().id()) {
+	case LogicalTypeId::INVALID:
+	case LogicalTypeId::SQLNULL: {
+		scalar->mutable_value()->set_null_value(google::protobuf::NULL_VALUE);
+		return scalar;
+	}
+	case LogicalTypeId::BOOLEAN: {
+		auto boolean = new vortex::dtype::Bool();
```

Comment:
Does this need to be heap allocated? And where is this freed.

Reply:
removed